File: algorithms/easy/find_the_k_or_of_an_array.dart
import 'dart:math';

class Solution {
  int findKOr(List<int> nums, int k) {
    int len = 31;
    final mp = <int, int>{};
    for (final num in nums) {
      for (int i = 0; i <= len; i++) {
        final t = pow(2, i).toInt();
        if (num & t == t) {
          mp[i] = (mp[i] ?? 0) + 1;
        }
      }
    }
    int ans = 0;
    for (final e in mp.entries) {
      if (e.value >= k) {
        ans += pow(2, e.key).toInt();
      }
    }
    return ans;
  }
}

File: algorithms/easy/find_the_k_or_of_an_array.ex
defmodule Solution do
  use Bitwise

  @spec find_k_or(nums :: [integer], k :: integer) :: integer
  def find_k_or(nums, k) do
    0..31
    |> Enum.reduce(0, fn i, ans ->
      t = 2 ** i

      nums
      |> Enum.reduce(0, fn num, acc ->
        acc + if (num &&& t) == t, do: 1, else: 0
      end)
      |> then(fn n ->
        ans + if n >= k, do: t, else: 0
      end)
    end)
  end
end
